test-data/unit/check-inference.test

```diff
@@ -1585,6 +1585,145 @@ oo.update(d)
 reveal_type(oo) # N: Revealed type is 'collections.OrderedDict[builtins.int*, builtins.str*]'
 [builtins fixtures/dict.pyi]
 
+[case testInferAttributeInitializedToEmptyAndAssigned]
+class C:
+    def __init__(self) -> None:
+        self.a = []
+        if bool():
+            self.a = [1]
+reveal_type(C().a)  # N: Revealed type is 'builtins.list[builtins.int*]'
+[builtins fixtures/list.pyi]
+
+[case testInferAttributeInitializedToEmptyAndAppended]
+class C:
+    def __init__(self) -> None:
+        self.a = []
+        if bool():
+            self.a.append(1)
+reveal_type(C().a)  # N: Revealed type is 'builtins.list[builtins.int]'
+[builtins fixtures/list.pyi]
+
+[case testInferAttributeInitializedToEmptyAndAssignedItem]
+class C:
+    def __init__(self) -> None:
+        self.a = {}
+        if bool():
+            self.a[0] = 'yes'
+reveal_type(C().a)  # N: Revealed type is 'builtins.dict[builtins.int, builtins.str]'
+[builtins fixtures/dict.pyi]
+
+[case testInferAttributeInitializedToNoneAndAssigned]
+# flags: --strict-optional
+class C:
+    def __init__(self) -> None:
+        self.a = None
+        if bool():
+            self.a = 1
+reveal_type(C().a)  # N: Revealed type is 'Union[builtins.int, None]'
+
+[case testInferAttributeInitializedToEmptyAndAssignedOtherMethod]
+class C:
+    def __init__(self) -> None:
+        self.a = []  # E: Need type annotation for 'a' (hint: "a: List[<type>] = ...")
+    def meth(self) -> None:
+        self.a = [1]
+reveal_type(C().a)  # N: Revealed type is 'builtins.list[Any]'
+[builtins fixtures/list.pyi]
+
+[case testInferAttributeInitializedToEmptyAndAppendedOtherMethod]
+class C:
+    def __init__(self) -> None:
+        self.a = []  # E: Need type annotation for 'a' (hint: "a: List[<type>] = ...")
+    def meth(self) -> None:
+        self.a.append(1)
+reveal_type(C().a)  # N: Revealed type is 'builtins.list[Any]'
+[builtins fixtures/list.pyi]
+
+[case testInferAttributeInitializedToEmptyAndAssignedItemOtherMethod]
+class C:
+    def __init__(self) -> None:
+        self.a = {}  # E: Need type annotation for 'a' (hint: "a: Dict[<type>, <type>] = ...")
+    def meth(self) -> None:
+        self.a[0] = 'yes'
+reveal_type(C().a)  # N: Revealed type is 'builtins.dict[Any, Any]'
+[builtins fixtures/dict.pyi]
+
+[case testInferAttributeInitializedToNoneAndAssignedOtherMethod]
+# flags: --strict-optional
+class C:
+    def __init__(self) -> None:
+        self.a = None
+    def meth(self) -> None:
+        self.a = 1  # E: Incompatible types in assignment (expression has type "int", variable has type "None")
+reveal_type(C().a)  # N: Revealed type is 'None'
+
+[case testInferAttributeInitializedToEmptyAndAssignedClassBody]
+class C:
+    a = []
+    def __init__(self) -> None:
+        self.a = [1]
```

Comment:
I'd rather not add support for this, since I want to deprecate not using `--local-partial-types` in the not too distant future (to make daemon and non-daemon modes uniform). If we further increase the semantic differences of `--local-partial-types`, the deprecation is going to be more painful.